packages/optimistic-oracle/test/keeper.js

```diff
@@ -0,0 +1,305 @@
+      }
+
+      // Check for the successful INFO log emitted by the keeper.
+      assert.equal(lastSpyLogLevel(spy), "info");
+      assert.isTrue(spyLogIncludes(spy, -1, "Proposed price"));
+
+      // After one run, the pricefeed classes should all be cached in the keerp bot's state:
+      for (let i = 0; i < identifiersToTest.length; i++) {
+        assert.isTrue(keeper.priceFeedCache[web3.utils.hexToUtf8(identifiersToTest[i])] instanceof PriceFeedMockScaled);
+      }
+    });
+
+    it("Submitting another price request for already used identifier uses cached price feed", async function() {
+      // Submit another price request (with a different timestamp) for an identifier that is already used.
+      await optimisticRequester.requestPrice(
+        identifiersToTest[0],
+        requestTime + 1,
+        "0x",
+        collateralCurrenciesForIdentifier[0].address,
+        0
+      );
+      // Update the keeper since the OO state has changed
+      await keeper.update();
+
+      // Submit all of the proposals now for all pending price requests.
+      await keeper.sendProposals();
+
+      // There should be one call for each unique collateral currency and one for each price request.
+      const expectedCallCount = collateralCurrenciesForIdentifier.length + identifiersToTest.length + 1;
+      assert.equal(expectedCallCount, spy.callCount);
+
+      // Check that only 1 price feed is cached for each unique identifier.
+      assert.equal(Object.keys(keeper.priceFeedCache).length, identifiersToTest.length);
+    });
+  });
+
+  it("Skip price requests with identifiers that keeper cannot construct a price feed for", async function() {
+    // Request a valid identifier but set an invalid price feed config:
+    await optimisticRequester.requestPrice(
+      identifiersToTest[0],
+      requestTime,
+      "0x",
+      collateralCurrenciesForIdentifier[0].address,
+      0
+    );
+
+    // This pricefeed config will cause the Keeper to fail to construct a price feed because the
+    // PriceFeedMock type requires a `currentPrice` and a `historicalPrice` to be specified, which are missing
+    // here (and also not included in the DefaultPriceFeedConfig for the tested identifiers).
+    let invalidPriceFeedConfig = {};
+    keeper = new OptimisticOracleKeeper({
+      logger: spyLogger,
+      optimisticOracleClient: client,
+      gasEstimator,
+      account: botRunner,
+      defaultPriceFeedConfig: invalidPriceFeedConfig
+    });
+    await keeper.update();
+    await keeper.sendProposals();
+
+    // Check for the specific failure log emitted by the keeper.
+    assert.equal(lastSpyLogLevel(spy), "error");
+    assert.isTrue(spyLogIncludes(spy, -1, "Failed to construct a PriceFeed for price request"));
+  });
+
+  it("Skip price requests with historical prices that keeper fails to fetch", async function() {
+    // Request a valid identifier but set an invalid price feed config.
```

Comment:
This is worded in a way that's a bit confusing, since it sounds the same as the earlier invalid price feed config. Maybe something like:

`// Request a valid identifier that is getting bad data from the data source`

Reply:
Great point! fixed